### Synthetic Data Backtest

Most of the functions below can be found under:

* Tool/backtest 
* Sample_data/make_data

A quick slideshow: [Seven Sins of Backtesting](https://newyork.qwafafew.org/wp-content/uploads/sites/4/2015/10/Luo_20150128.pdf)

The below is a code snippet for trading rules optimization.

The origin exercise required to use real-HFT data to deduce a certain outcome.. which I don't have.

Hence, I will just replace this exercise with multiprocessing methods (Parallelization!).

If you are keen on generating synthetic data for your research, copy the code snippets [Generate synthetic raw data](https://gist.github.com/boyboi86/5e00faf48f60abfdbe838fbdee269471) in my gist.

Contact: boyboi86@gmail.com

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import research as rs
from random import gauss
from itertools import product

%matplotlib inline

Num of CPU core:  4
Machine info:  Windows-10-10.0.18362-SP0
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]
Numpy 1.18.5
Pandas 1.0.4


In [2]:
dict0 = {'a':['1', '2'], 'b':['+', '*'], 'c':['!','A']}
for a in dict0['a']:
    for b in dict0['b']:
        for c in dict0['c']:
            print({'a':a, 'b':b, 'c':c})

{'a': '1', 'b': '+', 'c': '!'}
{'a': '1', 'b': '+', 'c': 'A'}
{'a': '1', 'b': '*', 'c': '!'}
{'a': '1', 'b': '*', 'c': 'A'}
{'a': '2', 'b': '+', 'c': '!'}
{'a': '2', 'b': '+', 'c': 'A'}
{'a': '2', 'b': '*', 'c': '!'}
{'a': '2', 'b': '*', 'c': 'A'}


**Note**

The above is scalar operation. Notice loop after loop after loop.

The below is vectorization operation, which is what we will do for Monte-Carlos.

The result is actually the same.

In [3]:
jobs = (dict(zip(dict0, i )) for i in product(*dict0.values()))
for i in jobs: print(i)

{'a': '1', 'b': '+', 'c': '!'}
{'a': '1', 'b': '+', 'c': 'A'}
{'a': '1', 'b': '*', 'c': '!'}
{'a': '1', 'b': '*', 'c': 'A'}
{'a': '2', 'b': '+', 'c': '!'}
{'a': '2', 'b': '+', 'c': 'A'}
{'a': '2', 'b': '*', 'c': '!'}
{'a': '2', 'b': '*', 'c': 'A'}


**Note**

Given the parameters below. It should look like the below.

In [4]:
forecast = [10,5,0,-5,-10]
half_life = [5,10,25,50,100]
loss_range = profit_range = np.linspace(.5, 10, 20)
sigma = [1]
max_period = half_life[-1]
n_run = 100

coeff = dict({'forecast': forecast,
              'half_life': half_life,
              'sigma':sigma,
              'profit_range': profit_range,
              'loss_range':loss_range})
jobs = list(dict(zip(coeff, idx)) for idx in product(*coeff.values()))

print(len(jobs)) #forecast * half_life * loss * profit * sigma
out = pd.DataFrame(jobs)
out

10000


forecast  half_life  sigma  profit_range  loss_range
0           10          5      1           0.5         0.5
1           10          5      1           0.5         1.0
2           10          5      1           0.5         1.5
3           10          5      1           0.5         2.0
4           10          5      1           0.5         2.5
...        ...        ...    ...           ...         ...
9995       -10        100      1          10.0         8.0
9996       -10        100      1          10.0         8.5
9997       -10        100      1          10.0         9.0
9998       -10        100      1          10.0         9.5
9999       -10        100      1          10.0        10.0

[10000 rows x 5 columns]

**Note**

The below is the end product for a non-parallized Monte-Carlos Simulation, for optimizied Trading Rule Algorithm.

Primarily we will use it to create scenarios for our models to be backtest with.

There is another algorithm which MLfinlab has implemented, you may wish to read that as well.

[Backtesting by Campbell and Yan [2015]](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2345489)

In [5]:
def _opt_tr(coeff: dict = None,
            n_run: float = 100,
            max_period: int = 100,
            seed: int = 0):
    
    phi, output1 = 2 ** (-1/ coeff['half_life']), []
    for iter_ in range(int(n_run)):
        p, hold_period, count = seed, 0, 0
        while True:
            p = (1 - phi) * coeff['forecast'] + phi * p + coeff['sigma'] * gauss(0,1)
            cp = p - seed
            hold_period += 1
            if coeff['profit_range'] < cp or -coeff['loss_range'] > cp or hold_period > max_period:
                output1.append(cp)
                break
    mean, std = np.mean(output1), np.std(output1)
    return {'profit': coeff['profit_range'], 'loss': coeff['loss_range'], 'mean': mean, 'std': std, 'sharpe': mean/std}

def opt_tr(profit_range: list = np.linspace(.5, 10, 20),
          loss_range: list = np.linspace(.5, 10, 20),
          sigma: list = [1],
          n_run: float = 100,
          max_period: int = 100,
          forecast: list = [10,5,0,-5,-10], 
          half_life: list = [5,10,25,50,100],
          num_threads: int = 1):
    
    if profit_range is None:
        profit_range = np.linspace(0,10,21)
        
    if loss_range is None:
        loss_range = np.linspace(0,10,21)
        
    _coeff_list = []
    _coeff = dict({'forecast': forecast,
                   'half_life': half_life,
                   'sigma':sigma,
                   'profit_range': profit_range,
                   'loss_range':loss_range})
    
    _coeff = list(dict(zip(_coeff, idx )) for idx in product(*_coeff.values()))
    out = []
    for coeff in _coeff:
        jobs = [{'func': _opt_tr,
                   'coeff': coeff,
                   'n_run': n_run, 
                   'max_period': max_period}]
        
        if num_threads == 1:
            out_ = rs.process_jobs_(jobs)
            out.append(out_)
        else:
            out_ = rs.process_jobs(jobs, num_threads = num_threads)
            out.append(out_)

    return out

df = opt_tr(profit_range = profit_range,
              loss_range = loss_range,
              sigma = sigma,
              n_run = n_run,
              max_period = max_period,
              forecast = forecast, 
              half_life = half_life,
              num_threads = 1)

2020-06-17 14:41:28.739905 done after 0.00 mins
2020-06-17 14:41:28.741900 done after 0.00 mins
2020-06-17 14:41:28.742898 done after 0.00 mins
2020-06-17 14:41:28.744894 done after 0.00 mins
2020-06-17 14:41:28.745889 done after 0.00 mins
2020-06-17 14:41:28.746887 done after 0.00 mins
2020-06-17 14:41:28.747883 done after 0.00 mins
2020-06-17 14:41:28.749879 done after 0.00 mins
2020-06-17 14:41:28.750875 done after 0.00 mins
2020-06-17 14:41:28.751873 done after 0.00 mins
2020-06-17 14:41:28.752870 done after 0.00 mins
2020-06-17 14:41:28.754865 done after 0.00 mins
2020-06-17 14:41:28.756001 done after 0.00 mins
2020-06-17 14:41:28.756001 done after 0.00 mins
2020-06-17 14:41:28.756001 done after 0.00 mins
2020-06-17 14:41:28.756001 done after 0.00 mins
2020-06-17 14:41:28.756001 done after 0.00 mins
2020-06-17 14:41:28.756001 done after 0.00 mins
2020-06-17 14:41:28.764031 done after 0.00 mins
2020-06-17 14:41:28.764031 done after 0.00 mins
2020-06-17 14:41:28.766057 done after 0.

2020-06-17 14:41:29.527935 done after 0.00 mins
2020-06-17 14:41:29.535974 done after 0.00 mins
2020-06-17 14:41:29.538002 done after 0.00 mins
2020-06-17 14:41:29.538002 done after 0.00 mins
2020-06-17 14:41:29.546040 done after 0.00 mins
2020-06-17 14:41:29.548069 done after 0.00 mins
2020-06-17 14:41:29.548069 done after 0.00 mins
2020-06-17 14:41:29.588342 done after 0.00 mins
2020-06-17 14:41:29.588342 done after 0.00 mins
2020-06-17 14:41:29.595040 done after 0.00 mins
2020-06-17 14:41:29.598032 done after 0.00 mins
2020-06-17 14:41:29.601024 done after 0.00 mins
2020-06-17 14:41:29.604016 done after 0.00 mins
2020-06-17 14:41:29.608006 done after 0.00 mins
2020-06-17 14:41:29.612994 done after 0.00 mins
2020-06-17 14:41:29.616984 done after 0.00 mins
2020-06-17 14:41:29.622237 done after 0.00 mins
2020-06-17 14:41:29.659086 done after 0.00 mins
2020-06-17 14:41:29.667120 done after 0.00 mins
2020-06-17 14:41:29.669153 done after 0.00 mins
2020-06-17 14:41:29.669153 done after 0.

2020-06-17 14:41:31.586821 done after 0.00 mins
2020-06-17 14:41:31.592380 done after 0.00 mins
2020-06-17 14:41:31.602464 done after 0.00 mins
2020-06-17 14:41:31.612528 done after 0.00 mins
2020-06-17 14:41:31.660900 done after 0.00 mins
2020-06-17 14:41:31.671104 done after 0.00 mins
2020-06-17 14:41:31.673240 done after 0.00 mins
2020-06-17 14:41:31.683564 done after 0.00 mins
2020-06-17 14:41:31.724143 done after 0.00 mins
2020-06-17 14:41:31.732265 done after 0.00 mins
2020-06-17 14:41:31.742368 done after 0.00 mins
2020-06-17 14:41:31.752433 done after 0.00 mins
2020-06-17 14:41:31.772573 done after 0.00 mins
2020-06-17 14:41:31.782637 done after 0.00 mins
2020-06-17 14:41:31.792912 done after 0.00 mins
2020-06-17 14:41:31.836663 done after 0.00 mins
2020-06-17 14:41:31.846882 done after 0.00 mins
2020-06-17 14:41:31.856856 done after 0.00 mins
2020-06-17 14:41:31.864834 done after 0.00 mins
2020-06-17 14:41:31.886906 done after 0.00 mins
2020-06-17 14:41:31.896966 done after 0.

2020-06-17 14:41:33.201956 done after 0.00 mins
2020-06-17 14:41:33.201956 done after 0.00 mins
2020-06-17 14:41:33.212064 done after 0.00 mins
2020-06-17 14:41:33.212064 done after 0.00 mins
2020-06-17 14:41:33.222134 done after 0.00 mins
2020-06-17 14:41:33.222134 done after 0.00 mins
2020-06-17 14:41:33.272963 done after 0.00 mins
2020-06-17 14:41:33.272963 done after 0.00 mins
2020-06-17 14:41:33.281002 done after 0.00 mins
2020-06-17 14:41:33.283030 done after 0.00 mins
2020-06-17 14:41:33.290958 done after 0.00 mins
2020-06-17 14:41:33.296942 done after 0.00 mins
2020-06-17 14:41:33.301929 done after 0.00 mins
2020-06-17 14:41:33.340935 done after 0.00 mins
2020-06-17 14:41:33.342961 done after 0.00 mins
2020-06-17 14:41:33.350991 done after 0.00 mins
2020-06-17 14:41:33.353042 done after 0.00 mins
2020-06-17 14:41:33.353042 done after 0.00 mins
2020-06-17 14:41:33.363124 done after 0.00 mins
2020-06-17 14:41:33.363124 done after 0.00 mins
2020-06-17 14:41:33.373188 done after 0.

2020-06-17 14:41:35.946711 done after 0.00 mins
2020-06-17 14:41:35.956779 done after 0.00 mins
2020-06-17 14:41:35.966887 done after 0.00 mins
2020-06-17 14:41:35.976949 done after 0.00 mins
2020-06-17 14:41:36.027579 done after 0.00 mins
2020-06-17 14:41:36.045709 done after 0.00 mins
2020-06-17 14:41:36.093386 done after 0.00 mins
2020-06-17 14:41:36.109348 done after 0.00 mins
2020-06-17 14:41:36.123306 done after 0.00 mins
2020-06-17 14:41:36.154895 done after 0.00 mins
2020-06-17 14:41:36.166491 done after 0.00 mins
2020-06-17 14:41:36.178635 done after 0.00 mins
2020-06-17 14:41:36.240011 done after 0.00 mins
2020-06-17 14:41:36.250093 done after 0.00 mins
2020-06-17 14:41:36.260267 done after 0.00 mins
2020-06-17 14:41:36.320986 done after 0.00 mins
2020-06-17 14:41:36.331490 done after 0.00 mins
2020-06-17 14:41:36.349834 done after 0.00 mins
2020-06-17 14:41:36.408013 done after 0.00 mins
2020-06-17 14:41:36.421983 done after 0.00 mins
2020-06-17 14:41:36.431948 done after 0.

2020-06-17 14:41:39.053946 done after 0.00 mins
2020-06-17 14:41:39.061993 done after 0.00 mins
2020-06-17 14:41:39.102554 done after 0.00 mins
2020-06-17 14:41:39.104620 done after 0.00 mins
2020-06-17 14:41:39.114689 done after 0.00 mins
2020-06-17 14:41:39.114689 done after 0.00 mins
2020-06-17 14:41:39.124851 done after 0.00 mins
2020-06-17 14:41:39.155314 done after 0.00 mins
2020-06-17 14:41:39.161321 done after 0.00 mins
2020-06-17 14:41:39.169302 done after 0.00 mins
2020-06-17 14:41:39.174287 done after 0.00 mins
2020-06-17 14:41:39.184259 done after 0.00 mins
2020-06-17 14:41:39.216175 done after 0.00 mins
2020-06-17 14:41:39.221222 done after 0.00 mins
2020-06-17 14:41:39.226291 done after 0.00 mins
2020-06-17 14:41:39.226291 done after 0.00 mins
2020-06-17 14:41:39.236364 done after 0.00 mins
2020-06-17 14:41:39.246436 done after 0.00 mins
2020-06-17 14:41:39.286842 done after 0.00 mins
2020-06-17 14:41:39.296911 done after 0.00 mins
2020-06-17 14:41:39.304940 done after 0.

2020-06-17 14:41:43.179300 done after 0.00 mins
2020-06-17 14:41:43.189357 done after 0.00 mins
2020-06-17 14:41:43.253681 done after 0.00 mins
2020-06-17 14:41:43.269304 done after 0.00 mins
2020-06-17 14:41:43.322446 done after 0.00 mins
2020-06-17 14:41:43.340592 done after 0.00 mins
2020-06-17 14:41:43.352693 done after 0.00 mins
2020-06-17 14:41:43.393286 done after 0.00 mins
2020-06-17 14:41:43.433640 done after 0.00 mins
2020-06-17 14:41:43.453782 done after 0.00 mins
2020-06-17 14:41:43.512289 done after 0.00 mins
2020-06-17 14:41:43.526405 done after 0.00 mins
2020-06-17 14:41:43.576035 done after 0.00 mins
2020-06-17 14:41:43.597248 done after 0.00 mins
2020-06-17 14:41:43.647668 done after 0.00 mins
2020-06-17 14:41:43.698683 done after 0.00 mins
2020-06-17 14:41:43.718626 done after 0.00 mins
2020-06-17 14:41:43.748704 done after 0.00 mins
2020-06-17 14:41:43.769140 done after 0.00 mins
2020-06-17 14:41:43.819684 done after 0.00 mins
2020-06-17 14:41:43.829755 done after 0.

2020-06-17 14:41:49.016316 done after 0.00 mins
2020-06-17 14:41:49.024354 done after 0.00 mins
2020-06-17 14:41:49.026381 done after 0.00 mins
2020-06-17 14:41:49.036510 done after 0.00 mins
2020-06-17 14:41:49.044546 done after 0.00 mins
2020-06-17 14:41:49.077528 done after 0.00 mins
2020-06-17 14:41:49.087616 done after 0.00 mins
2020-06-17 14:41:49.103212 done after 0.00 mins
2020-06-17 14:41:49.147878 done after 0.00 mins
2020-06-17 14:41:49.157140 done after 0.00 mins
2020-06-17 14:41:49.159172 done after 0.00 mins
2020-06-17 14:41:49.169249 done after 0.00 mins
2020-06-17 14:41:49.179322 done after 0.00 mins
2020-06-17 14:41:49.210170 done after 0.00 mins
2020-06-17 14:41:49.220226 done after 0.00 mins
2020-06-17 14:41:49.228293 done after 0.00 mins
2020-06-17 14:41:49.230322 done after 0.00 mins
2020-06-17 14:41:49.230322 done after 0.00 mins
2020-06-17 14:41:49.238354 done after 0.00 mins
2020-06-17 14:41:49.240391 done after 0.00 mins
2020-06-17 14:41:49.290868 done after 0.

2020-06-17 14:41:52.660804 done after 0.00 mins
2020-06-17 14:41:52.729719 done after 0.00 mins
2020-06-17 14:41:52.736701 done after 0.00 mins
2020-06-17 14:41:52.766619 done after 0.00 mins
2020-06-17 14:41:52.779585 done after 0.00 mins
2020-06-17 14:41:52.793547 done after 0.00 mins
2020-06-17 14:41:52.852389 done after 0.00 mins
2020-06-17 14:41:52.901356 done after 0.00 mins
2020-06-17 14:41:52.919442 done after 0.00 mins
2020-06-17 14:41:52.971877 done after 0.00 mins
2020-06-17 14:41:52.992003 done after 0.00 mins
2020-06-17 14:41:53.052396 done after 0.00 mins
2020-06-17 14:41:53.082702 done after 0.00 mins
2020-06-17 14:41:53.181339 done after 0.00 mins
2020-06-17 14:41:53.259984 done after 0.00 mins
2020-06-17 14:41:53.285908 done after 0.00 mins
2020-06-17 14:41:53.493503 done after 0.00 mins
2020-06-17 14:41:53.660115 done after 0.00 mins
2020-06-17 14:41:53.764834 done after 0.00 mins
2020-06-17 14:41:54.198185 done after 0.01 mins
2020-06-17 14:41:54.227108 done after 0.

2020-06-17 14:42:02.918174 done after 0.00 mins
2020-06-17 14:42:02.926153 done after 0.00 mins
2020-06-17 14:42:02.928146 done after 0.00 mins
2020-06-17 14:42:02.931139 done after 0.00 mins
2020-06-17 14:42:02.977568 done after 0.00 mins
2020-06-17 14:42:02.979596 done after 0.00 mins
2020-06-17 14:42:02.987627 done after 0.00 mins
2020-06-17 14:42:02.989655 done after 0.00 mins
2020-06-17 14:42:02.989655 done after 0.00 mins
2020-06-17 14:42:02.999773 done after 0.00 mins
2020-06-17 14:42:03.009845 done after 0.00 mins
2020-06-17 14:42:03.009845 done after 0.00 mins
2020-06-17 14:42:03.048124 done after 0.00 mins
2020-06-17 14:42:03.050157 done after 0.00 mins
2020-06-17 14:42:03.060251 done after 0.00 mins
2020-06-17 14:42:03.070312 done after 0.00 mins
2020-06-17 14:42:03.120768 done after 0.00 mins
2020-06-17 14:42:03.130828 done after 0.00 mins
2020-06-17 14:42:03.140965 done after 0.00 mins
2020-06-17 14:42:03.202325 done after 0.00 mins
2020-06-17 14:42:03.212384 done after 0.

2020-06-17 14:42:07.120583 done after 0.00 mins
2020-06-17 14:42:07.181421 done after 0.00 mins
2020-06-17 14:42:07.243759 done after 0.00 mins
2020-06-17 14:42:07.273925 done after 0.00 mins
2020-06-17 14:42:07.334504 done after 0.00 mins
2020-06-17 14:42:07.390896 done after 0.00 mins
2020-06-17 14:42:07.397878 done after 0.00 mins
2020-06-17 14:42:07.402866 done after 0.00 mins
2020-06-17 14:42:07.447745 done after 0.00 mins
2020-06-17 14:42:07.455723 done after 0.00 mins
2020-06-17 14:42:07.464797 done after 0.00 mins
2020-06-17 14:42:07.521044 done after 0.00 mins
2020-06-17 14:42:07.541984 done after 0.00 mins
2020-06-17 14:42:07.608803 done after 0.00 mins
2020-06-17 14:42:07.628751 done after 0.00 mins
2020-06-17 14:42:07.683603 done after 0.00 mins
2020-06-17 14:42:07.748430 done after 0.00 mins
2020-06-17 14:42:07.777352 done after 0.00 mins
2020-06-17 14:42:07.873099 done after 0.00 mins
2020-06-17 14:42:07.948626 done after 0.00 mins
2020-06-17 14:42:08.014449 done after 0.

2020-06-17 14:42:15.901066 done after 0.00 mins
2020-06-17 14:42:15.961903 done after 0.00 mins
2020-06-17 14:42:16.026353 done after 0.00 mins
2020-06-17 14:42:16.082333 done after 0.00 mins
2020-06-17 14:42:16.115134 done after 0.00 mins
2020-06-17 14:42:16.242132 done after 0.00 mins
2020-06-17 14:42:16.319060 done after 0.00 mins
2020-06-17 14:42:16.391865 done after 0.00 mins
2020-06-17 14:42:16.470188 done after 0.00 mins
2020-06-17 14:42:16.567423 done after 0.00 mins
2020-06-17 14:42:16.649208 done after 0.00 mins
2020-06-17 14:42:16.726422 done after 0.00 mins
2020-06-17 14:42:16.812099 done after 0.00 mins
2020-06-17 14:42:16.948008 done after 0.00 mins
2020-06-17 14:42:16.948008 done after 0.00 mins
2020-06-17 14:42:16.948008 done after 0.00 mins
2020-06-17 14:42:16.956928 done after 0.00 mins
2020-06-17 14:42:16.958923 done after 0.00 mins
2020-06-17 14:42:16.961915 done after 0.00 mins
2020-06-17 14:42:16.963909 done after 0.00 mins
2020-06-17 14:42:16.964966 done after 0.

2020-06-17 14:42:18.309789 done after 0.00 mins
2020-06-17 14:42:18.316771 done after 0.00 mins
2020-06-17 14:42:18.324750 done after 0.00 mins
2020-06-17 14:42:18.332729 done after 0.00 mins
2020-06-17 14:42:18.336905 done after 0.00 mins
2020-06-17 14:42:18.396328 done after 0.00 mins
2020-06-17 14:42:18.398356 done after 0.00 mins
2020-06-17 14:42:18.408423 done after 0.00 mins
2020-06-17 14:42:18.418560 done after 0.00 mins
2020-06-17 14:42:18.418560 done after 0.00 mins
2020-06-17 14:42:18.469215 done after 0.00 mins
2020-06-17 14:42:18.469215 done after 0.00 mins
2020-06-17 14:42:18.479319 done after 0.00 mins
2020-06-17 14:42:18.489851 done after 0.00 mins
2020-06-17 14:42:18.497937 done after 0.00 mins
2020-06-17 14:42:18.520160 done after 0.00 mins
2020-06-17 14:42:18.530226 done after 0.00 mins
2020-06-17 14:42:18.540326 done after 0.00 mins
2020-06-17 14:42:18.548360 done after 0.00 mins
2020-06-17 14:42:18.582752 done after 0.00 mins
2020-06-17 14:42:18.590730 done after 0.

2020-06-17 14:42:24.176603 done after 0.00 mins
2020-06-17 14:42:24.247120 done after 0.00 mins
2020-06-17 14:42:24.318524 done after 0.00 mins
2020-06-17 14:42:24.389799 done after 0.00 mins
2020-06-17 14:42:24.463778 done after 0.00 mins
2020-06-17 14:42:24.519693 done after 0.00 mins
2020-06-17 14:42:24.582557 done after 0.00 mins
2020-06-17 14:42:24.652990 done after 0.00 mins
2020-06-17 14:42:24.715075 done after 0.00 mins
2020-06-17 14:42:24.755557 done after 0.00 mins
2020-06-17 14:42:24.826097 done after 0.00 mins
2020-06-17 14:42:24.941249 done after 0.00 mins
2020-06-17 14:42:25.009080 done after 0.00 mins
2020-06-17 14:42:25.069609 done after 0.00 mins
2020-06-17 14:42:25.141038 done after 0.00 mins
2020-06-17 14:42:25.212583 done after 0.00 mins
2020-06-17 14:42:25.342336 done after 0.00 mins
2020-06-17 14:42:25.433059 done after 0.00 mins
2020-06-17 14:42:25.562704 done after 0.00 mins
2020-06-17 14:42:25.635319 done after 0.00 mins
2020-06-17 14:42:25.696299 done after 0.

2020-06-17 14:42:30.083872 done after 0.00 mins
2020-06-17 14:42:30.093968 done after 0.00 mins
2020-06-17 14:42:30.104038 done after 0.00 mins
2020-06-17 14:42:30.104038 done after 0.00 mins
2020-06-17 14:42:30.144403 done after 0.00 mins
2020-06-17 14:42:30.152475 done after 0.00 mins
2020-06-17 14:42:30.155080 done after 0.00 mins
2020-06-17 14:42:30.164651 done after 0.00 mins
2020-06-17 14:42:30.172777 done after 0.00 mins
2020-06-17 14:42:30.179758 done after 0.00 mins
2020-06-17 14:42:30.224638 done after 0.00 mins
2020-06-17 14:42:30.239600 done after 0.00 mins
2020-06-17 14:42:30.247576 done after 0.00 mins
2020-06-17 14:42:30.259552 done after 0.00 mins
2020-06-17 14:42:30.298445 done after 0.00 mins
2020-06-17 14:42:30.307419 done after 0.00 mins
2020-06-17 14:42:30.316394 done after 0.00 mins
2020-06-17 14:42:30.361480 done after 0.00 mins
2020-06-17 14:42:30.370457 done after 0.00 mins
2020-06-17 14:42:30.378435 done after 0.00 mins
2020-06-17 14:42:30.386414 done after 0.

2020-06-17 14:42:35.343805 done after 0.00 mins
2020-06-17 14:42:35.432803 done after 0.00 mins
2020-06-17 14:42:35.523692 done after 0.00 mins
2020-06-17 14:42:35.547628 done after 0.00 mins
2020-06-17 14:42:35.612024 done after 0.00 mins
2020-06-17 14:42:35.678187 done after 0.00 mins
2020-06-17 14:42:35.749946 done after 0.00 mins
2020-06-17 14:42:35.774877 done after 0.00 mins
2020-06-17 14:42:35.822751 done after 0.00 mins
2020-06-17 14:42:35.861173 done after 0.00 mins
2020-06-17 14:42:35.922134 done after 0.00 mins
2020-06-17 14:42:35.992824 done after 0.00 mins
2020-06-17 14:42:36.021063 done after 0.00 mins
2020-06-17 14:42:36.072855 done after 0.00 mins
2020-06-17 14:42:36.088811 done after 0.00 mins
2020-06-17 14:42:36.132160 done after 0.00 mins
2020-06-17 14:42:36.186474 done after 0.00 mins
2020-06-17 14:42:36.206368 done after 0.00 mins
2020-06-17 14:42:36.267031 done after 0.00 mins
2020-06-17 14:42:36.357861 done after 0.00 mins
2020-06-17 14:42:36.460745 done after 0.

2020-06-17 14:42:43.899065 done after 0.00 mins
2020-06-17 14:42:43.902057 done after 0.00 mins
2020-06-17 14:42:43.906046 done after 0.00 mins
2020-06-17 14:42:43.910034 done after 0.00 mins
2020-06-17 14:42:43.914024 done after 0.00 mins
2020-06-17 14:42:43.919012 done after 0.00 mins
2020-06-17 14:42:43.924996 done after 0.00 mins
2020-06-17 14:42:43.977651 done after 0.00 mins
2020-06-17 14:42:43.984056 done after 0.00 mins
2020-06-17 14:42:43.994380 done after 0.00 mins
2020-06-17 14:42:43.998425 done after 0.00 mins
2020-06-17 14:42:44.011830 done after 0.00 mins
2020-06-17 14:42:44.065413 done after 0.00 mins
2020-06-17 14:42:44.079021 done after 0.00 mins
2020-06-17 14:42:44.086136 done after 0.00 mins
2020-06-17 14:42:44.123470 done after 0.00 mins
2020-06-17 14:42:44.131547 done after 0.00 mins
2020-06-17 14:42:44.142077 done after 0.00 mins
2020-06-17 14:42:44.149059 done after 0.00 mins
2020-06-17 14:42:44.195935 done after 0.00 mins
2020-06-17 14:42:44.199922 done after 0.

2020-06-17 14:42:48.297033 done after 0.00 mins
2020-06-17 14:42:48.436774 done after 0.00 mins
2020-06-17 14:42:48.534562 done after 0.00 mins
2020-06-17 14:42:48.561490 done after 0.00 mins
2020-06-17 14:42:48.616349 done after 0.00 mins
2020-06-17 14:42:48.676183 done after 0.00 mins
2020-06-17 14:42:48.734137 done after 0.00 mins
2020-06-17 14:42:48.836739 done after 0.00 mins
2020-06-17 14:42:48.896463 done after 0.00 mins
2020-06-17 14:42:48.896463 done after 0.00 mins
2020-06-17 14:42:48.906576 done after 0.00 mins
2020-06-17 14:42:48.924925 done after 0.00 mins
2020-06-17 14:42:48.957221 done after 0.00 mins
2020-06-17 14:42:48.975791 done after 0.00 mins
2020-06-17 14:42:49.051308 done after 0.00 mins
2020-06-17 14:42:49.070258 done after 0.00 mins
2020-06-17 14:42:49.129751 done after 0.00 mins
2020-06-17 14:42:49.180461 done after 0.00 mins
2020-06-17 14:42:49.241524 done after 0.00 mins
2020-06-17 14:42:49.261663 done after 0.00 mins
2020-06-17 14:42:49.341074 done after 0.

2020-06-17 14:42:58.805645 done after 0.00 mins
2020-06-17 14:42:58.815714 done after 0.00 mins
2020-06-17 14:42:58.815714 done after 0.00 mins
2020-06-17 14:42:58.825769 done after 0.00 mins
2020-06-17 14:42:58.825769 done after 0.00 mins
2020-06-17 14:42:58.835898 done after 0.00 mins
2020-06-17 14:42:58.889522 done after 0.00 mins
2020-06-17 14:42:58.896504 done after 0.00 mins
2020-06-17 14:42:58.898505 done after 0.00 mins
2020-06-17 14:42:58.901491 done after 0.00 mins
2020-06-17 14:42:58.904482 done after 0.00 mins
2020-06-17 14:42:58.907474 done after 0.00 mins
2020-06-17 14:42:58.910467 done after 0.00 mins
2020-06-17 14:42:58.915453 done after 0.00 mins
2020-06-17 14:42:58.936749 done after 0.00 mins
2020-06-17 14:42:58.946812 done after 0.00 mins
2020-06-17 14:42:58.946812 done after 0.00 mins
2020-06-17 14:42:58.956896 done after 0.00 mins
2020-06-17 14:42:58.966956 done after 0.00 mins
2020-06-17 14:42:58.966956 done after 0.00 mins
2020-06-17 14:42:59.020368 done after 0.

2020-06-17 14:43:02.882004 done after 0.00 mins
2020-06-17 14:43:02.902485 done after 0.00 mins
2020-06-17 14:43:02.974763 done after 0.00 mins
2020-06-17 14:43:03.032813 done after 0.00 mins
2020-06-17 14:43:03.055029 done after 0.00 mins
2020-06-17 14:43:03.095900 done after 0.00 mins
2020-06-17 14:43:03.116371 done after 0.00 mins
2020-06-17 14:43:03.170753 done after 0.00 mins
2020-06-17 14:43:03.206908 done after 0.00 mins
2020-06-17 14:43:03.248180 done after 0.00 mins
2020-06-17 14:43:03.278744 done after 0.00 mins
2020-06-17 14:43:03.349294 done after 0.00 mins
2020-06-17 14:43:03.410422 done after 0.00 mins
2020-06-17 14:43:03.418491 done after 0.00 mins
2020-06-17 14:43:03.425673 done after 0.00 mins
2020-06-17 14:43:03.450603 done after 0.00 mins
2020-06-17 14:43:03.462572 done after 0.00 mins
2020-06-17 14:43:03.470426 done after 0.00 mins
2020-06-17 14:43:03.520821 done after 0.00 mins
2020-06-17 14:43:03.530939 done after 0.00 mins
2020-06-17 14:43:03.551419 done after 0.

2020-06-17 14:43:11.210214 done after 0.00 mins
2020-06-17 14:43:11.219190 done after 0.00 mins
2020-06-17 14:43:11.271052 done after 0.00 mins
2020-06-17 14:43:11.287009 done after 0.00 mins
2020-06-17 14:43:11.337873 done after 0.00 mins
2020-06-17 14:43:11.355824 done after 0.00 mins
2020-06-17 14:43:11.421963 done after 0.00 mins
2020-06-17 14:43:11.444027 done after 0.00 mins
2020-06-17 14:43:11.505096 done after 0.00 mins
2020-06-17 14:43:11.603877 done after 0.00 mins
2020-06-17 14:43:11.696811 done after 0.00 mins
2020-06-17 14:43:11.776809 done after 0.00 mins
2020-06-17 14:43:11.847365 done after 0.00 mins
2020-06-17 14:43:11.919376 done after 0.00 mins
2020-06-17 14:43:11.989256 done after 0.00 mins
2020-06-17 14:43:12.082861 done after 0.00 mins
2020-06-17 14:43:12.143757 done after 0.00 mins
2020-06-17 14:43:12.174026 done after 0.00 mins
2020-06-17 14:43:12.291963 done after 0.00 mins
2020-06-17 14:43:12.343450 done after 0.00 mins
2020-06-17 14:43:12.416383 done after 0.

2020-06-17 14:43:15.180624 done after 0.00 mins
2020-06-17 14:43:15.268277 done after 0.00 mins
2020-06-17 14:43:15.351286 done after 0.00 mins
2020-06-17 14:43:15.353358 done after 0.00 mins
2020-06-17 14:43:15.361399 done after 0.00 mins
2020-06-17 14:43:15.363427 done after 0.00 mins
2020-06-17 14:43:15.373488 done after 0.00 mins
2020-06-17 14:43:15.383615 done after 0.00 mins
2020-06-17 14:43:15.477276 done after 0.00 mins
2020-06-17 14:43:15.485986 done after 0.00 mins
2020-06-17 14:43:15.494798 done after 0.00 mins
2020-06-17 14:43:15.563483 done after 0.00 mins
2020-06-17 14:43:15.575569 done after 0.00 mins
2020-06-17 14:43:15.636759 done after 0.00 mins
2020-06-17 14:43:15.647111 done after 0.00 mins
2020-06-17 14:43:15.717767 done after 0.00 mins
2020-06-17 14:43:15.735881 done after 0.00 mins
2020-06-17 14:43:15.747985 done after 0.00 mins
2020-06-17 14:43:15.808525 done after 0.00 mins
2020-06-17 14:43:15.866325 done after 0.00 mins
2020-06-17 14:43:15.919800 done after 0.

2020-06-17 14:43:23.381390 done after 0.00 mins
2020-06-17 14:43:23.459557 done after 0.00 mins
2020-06-17 14:43:23.523653 done after 0.00 mins
2020-06-17 14:43:23.557563 done after 0.00 mins
2020-06-17 14:43:23.627374 done after 0.00 mins
2020-06-17 14:43:23.686195 done after 0.00 mins
2020-06-17 14:43:23.741499 done after 0.00 mins
2020-06-17 14:43:23.823683 done after 0.00 mins
2020-06-17 14:43:23.928370 done after 0.00 mins
2020-06-17 14:43:24.029130 done after 0.00 mins
2020-06-17 14:43:24.102677 done after 0.00 mins
2020-06-17 14:43:24.201413 done after 0.00 mins
2020-06-17 14:43:24.208936 done after 0.00 mins
2020-06-17 14:43:24.220854 done after 0.00 mins
2020-06-17 14:43:24.232817 done after 0.00 mins
2020-06-17 14:43:24.343636 done after 0.00 mins
2020-06-17 14:43:24.482558 done after 0.00 mins
2020-06-17 14:43:24.505497 done after 0.00 mins
2020-06-17 14:43:24.579651 done after 0.00 mins
2020-06-17 14:43:24.627524 done after 0.00 mins
2020-06-17 14:43:24.653576 done after 0.

2020-06-17 14:43:29.642162 done after 0.00 mins
2020-06-17 14:43:29.646151 done after 0.00 mins
2020-06-17 14:43:29.719337 done after 0.00 mins
2020-06-17 14:43:29.726319 done after 0.00 mins
2020-06-17 14:43:29.733299 done after 0.00 mins
2020-06-17 14:43:29.743272 done after 0.00 mins
2020-06-17 14:43:29.813085 done after 0.00 mins
2020-06-17 14:43:29.829047 done after 0.00 mins
2020-06-17 14:43:29.884894 done after 0.00 mins
2020-06-17 14:43:29.908135 done after 0.00 mins
2020-06-17 14:43:29.973960 done after 0.00 mins
2020-06-17 14:43:30.016845 done after 0.00 mins
2020-06-17 14:43:30.041776 done after 0.00 mins
2020-06-17 14:43:30.103612 done after 0.00 mins
2020-06-17 14:43:30.171172 done after 0.00 mins
2020-06-17 14:43:30.194739 done after 0.00 mins
2020-06-17 14:43:30.262556 done after 0.00 mins
2020-06-17 14:43:30.332369 done after 0.00 mins
2020-06-17 14:43:30.412158 done after 0.00 mins
2020-06-17 14:43:30.438090 done after 0.00 mins
2020-06-17 14:43:30.477981 done after 0.

2020-06-17 14:43:40.122398 done after 0.00 mins
2020-06-17 14:43:40.251884 done after 0.00 mins
2020-06-17 14:43:40.361179 done after 0.00 mins
2020-06-17 14:43:40.504012 done after 0.00 mins
2020-06-17 14:43:40.605364 done after 0.00 mins
2020-06-17 14:43:40.725322 done after 0.00 mins
2020-06-17 14:43:40.829017 done after 0.00 mins
2020-06-17 14:43:40.829017 done after 0.00 mins
2020-06-17 14:43:40.839108 done after 0.00 mins
2020-06-17 14:43:40.849203 done after 0.00 mins
2020-06-17 14:43:40.900085 done after 0.00 mins
2020-06-17 14:43:40.917039 done after 0.00 mins
2020-06-17 14:43:40.975882 done after 0.00 mins
2020-06-17 14:43:41.030203 done after 0.00 mins
2020-06-17 14:43:41.102165 done after 0.00 mins
2020-06-17 14:43:41.173637 done after 0.00 mins
2020-06-17 14:43:41.242416 done after 0.00 mins
2020-06-17 14:43:41.346548 done after 0.00 mins
2020-06-17 14:43:41.529602 done after 0.00 mins
2020-06-17 14:43:41.630519 done after 0.00 mins
2020-06-17 14:43:41.843683 done after 0.

2020-06-17 14:43:50.609941 done after 0.00 mins
2020-06-17 14:43:50.638271 done after 0.00 mins
2020-06-17 14:43:50.648245 done after 0.00 mins
2020-06-17 14:43:50.681156 done after 0.00 mins
2020-06-17 14:43:50.691282 done after 0.00 mins
2020-06-17 14:43:50.699320 done after 0.00 mins
2020-06-17 14:43:50.741800 done after 0.00 mins
2020-06-17 14:43:50.751931 done after 0.00 mins
2020-06-17 14:43:50.761999 done after 0.00 mins
2020-06-17 14:43:50.772071 done after 0.00 mins
2020-06-17 14:43:50.820809 done after 0.00 mins
2020-06-17 14:43:50.822837 done after 0.00 mins
2020-06-17 14:43:50.822837 done after 0.00 mins
2020-06-17 14:43:50.831065 done after 0.00 mins
2020-06-17 14:43:50.835249 done after 0.00 mins
2020-06-17 14:43:50.837370 done after 0.00 mins
2020-06-17 14:43:50.843443 done after 0.00 mins
2020-06-17 14:43:50.884944 done after 0.00 mins
2020-06-17 14:43:50.893922 done after 0.00 mins
2020-06-17 14:43:50.901900 done after 0.00 mins
2020-06-17 14:43:50.944313 done after 0.

2020-06-17 14:43:57.681519 done after 0.00 mins
2020-06-17 14:43:57.735498 done after 0.00 mins
2020-06-17 14:43:57.845735 done after 0.00 mins
2020-06-17 14:43:57.926315 done after 0.00 mins
2020-06-17 14:43:57.999037 done after 0.00 mins
2020-06-17 14:43:58.110855 done after 0.00 mins
2020-06-17 14:43:58.186241 done after 0.00 mins
2020-06-17 14:43:58.264371 done after 0.00 mins
2020-06-17 14:43:58.344745 done after 0.00 mins
2020-06-17 14:43:58.446879 done after 0.00 mins
2020-06-17 14:43:58.567372 done after 0.00 mins
2020-06-17 14:43:58.639298 done after 0.00 mins
2020-06-17 14:43:58.647343 done after 0.00 mins
2020-06-17 14:43:58.649371 done after 0.00 mins
2020-06-17 14:43:58.689823 done after 0.00 mins
2020-06-17 14:43:58.700058 done after 0.00 mins
2020-06-17 14:43:58.755633 done after 0.00 mins
2020-06-17 14:43:58.772591 done after 0.00 mins
2020-06-17 14:43:58.830273 done after 0.00 mins
2020-06-17 14:43:58.882494 done after 0.00 mins
2020-06-17 14:43:58.943094 done after 0.

2020-06-17 14:44:10.083147 done after 0.00 mins
2020-06-17 14:44:10.091179 done after 0.00 mins
2020-06-17 14:44:10.133869 done after 0.00 mins
2020-06-17 14:44:10.137820 done after 0.00 mins
2020-06-17 14:44:10.140810 done after 0.00 mins
2020-06-17 14:44:10.143802 done after 0.00 mins
2020-06-17 14:44:10.146794 done after 0.00 mins
2020-06-17 14:44:10.148789 done after 0.00 mins
2020-06-17 14:44:10.152779 done after 0.00 mins
2020-06-17 14:44:10.157012 done after 0.00 mins
2020-06-17 14:44:10.163750 done after 0.00 mins
2020-06-17 14:44:10.168736 done after 0.00 mins
2020-06-17 14:44:10.196662 done after 0.00 mins
2020-06-17 14:44:10.201648 done after 0.00 mins
2020-06-17 14:44:10.204641 done after 0.00 mins
2020-06-17 14:44:10.213719 done after 0.00 mins
2020-06-17 14:44:10.213719 done after 0.00 mins
2020-06-17 14:44:10.223856 done after 0.00 mins
2020-06-17 14:44:10.223856 done after 0.00 mins
2020-06-17 14:44:10.274904 done after 0.00 mins
2020-06-17 14:44:10.325732 done after 0.

2020-06-17 14:44:13.960964 done after 0.00 mins
2020-06-17 14:44:13.995157 done after 0.00 mins
2020-06-17 14:44:14.095876 done after 0.00 mins
2020-06-17 14:44:14.120115 done after 0.00 mins
2020-06-17 14:44:14.185597 done after 0.00 mins
2020-06-17 14:44:14.190583 done after 0.00 mins
2020-06-17 14:44:14.233471 done after 0.00 mins
2020-06-17 14:44:14.235401 done after 0.00 mins
2020-06-17 14:44:14.243457 done after 0.00 mins
2020-06-17 14:44:14.253521 done after 0.00 mins
2020-06-17 14:44:14.263585 done after 0.00 mins
2020-06-17 14:44:14.324255 done after 0.00 mins
2020-06-17 14:44:14.342418 done after 0.00 mins
2020-06-17 14:44:14.423632 done after 0.00 mins
2020-06-17 14:44:14.445791 done after 0.00 mins
2020-06-17 14:44:14.527703 done after 0.00 mins
2020-06-17 14:44:14.548030 done after 0.00 mins
2020-06-17 14:44:14.618663 done after 0.00 mins
2020-06-17 14:44:14.688760 done after 0.00 mins
2020-06-17 14:44:14.739137 done after 0.00 mins
2020-06-17 14:44:14.789482 done after 0.

2020-06-17 14:44:23.446110 done after 0.00 mins
2020-06-17 14:44:23.527709 done after 0.00 mins
2020-06-17 14:44:23.568485 done after 0.00 mins
2020-06-17 14:44:23.644830 done after 0.00 mins
2020-06-17 14:44:23.741159 done after 0.00 mins
2020-06-17 14:44:23.830007 done after 0.00 mins
2020-06-17 14:44:23.901055 done after 0.00 mins
2020-06-17 14:44:23.971954 done after 0.00 mins
2020-06-17 14:44:24.052651 done after 0.00 mins
2020-06-17 14:44:24.173816 done after 0.00 mins
2020-06-17 14:44:24.247700 done after 0.00 mins
2020-06-17 14:44:24.255652 done after 0.00 mins
2020-06-17 14:44:24.257676 done after 0.00 mins
2020-06-17 14:44:24.306401 done after 0.00 mins
2020-06-17 14:44:24.318512 done after 0.00 mins
2020-06-17 14:44:24.366947 done after 0.00 mins
2020-06-17 14:44:24.419480 done after 0.00 mins
2020-06-17 14:44:24.447683 done after 0.00 mins
2020-06-17 14:44:24.520533 done after 0.00 mins
2020-06-17 14:44:24.601629 done after 0.00 mins
2020-06-17 14:44:24.669399 done after 0.

2020-06-17 14:44:28.946335 done after 0.00 mins
2020-06-17 14:44:28.955312 done after 0.00 mins
2020-06-17 14:44:28.966794 done after 0.00 mins
2020-06-17 14:44:29.000703 done after 0.00 mins
2020-06-17 14:44:29.012896 done after 0.00 mins
2020-06-17 14:44:29.023127 done after 0.00 mins
2020-06-17 14:44:29.063895 done after 0.00 mins
2020-06-17 14:44:29.114730 done after 0.00 mins
2020-06-17 14:44:29.133020 done after 0.00 mins
2020-06-17 14:44:29.145150 done after 0.00 mins
2020-06-17 14:44:29.196090 done after 0.00 mins
2020-06-17 14:44:29.257679 done after 0.00 mins
2020-06-17 14:44:29.278624 done after 0.00 mins
2020-06-17 14:44:29.332479 done after 0.00 mins
2020-06-17 14:44:29.410830 done after 0.00 mins
2020-06-17 14:44:29.428963 done after 0.00 mins
2020-06-17 14:44:29.439023 done after 0.00 mins
2020-06-17 14:44:29.439023 done after 0.00 mins
2020-06-17 14:44:29.541938 done after 0.00 mins
2020-06-17 14:44:29.546923 done after 0.00 mins
2020-06-17 14:44:29.552214 done after 0.

2020-06-17 14:44:36.005162 done after 0.00 mins
2020-06-17 14:44:36.086286 done after 0.00 mins
2020-06-17 14:44:36.135727 done after 0.00 mins
2020-06-17 14:44:36.137808 done after 0.00 mins
2020-06-17 14:44:36.188597 done after 0.00 mins
2020-06-17 14:44:36.198783 done after 0.00 mins
2020-06-17 14:44:36.269669 done after 0.00 mins
2020-06-17 14:44:36.287807 done after 0.00 mins
2020-06-17 14:44:36.330215 done after 0.00 mins
2020-06-17 14:44:36.357196 done after 0.00 mins
2020-06-17 14:44:36.412044 done after 0.00 mins
2020-06-17 14:44:36.472289 done after 0.00 mins
2020-06-17 14:44:36.502547 done after 0.00 mins
2020-06-17 14:44:36.560967 done after 0.00 mins
2020-06-17 14:44:36.621543 done after 0.00 mins
2020-06-17 14:44:36.672912 done after 0.00 mins
2020-06-17 14:44:36.724530 done after 0.00 mins
2020-06-17 14:44:36.785408 done after 0.00 mins
2020-06-17 14:44:36.865191 done after 0.00 mins
2020-06-17 14:44:36.937205 done after 0.00 mins
2020-06-17 14:44:36.997942 done after 0.

2020-06-17 14:44:42.368790 done after 0.00 mins
2020-06-17 14:44:42.381114 done after 0.00 mins
2020-06-17 14:44:42.401645 done after 0.00 mins
2020-06-17 14:44:42.438158 done after 0.00 mins
2020-06-17 14:44:42.449217 done after 0.00 mins
2020-06-17 14:44:42.469240 done after 0.00 mins
2020-06-17 14:44:42.480191 done after 0.00 mins
2020-06-17 14:44:42.492314 done after 0.00 mins
2020-06-17 14:44:42.542713 done after 0.00 mins
2020-06-17 14:44:42.552788 done after 0.00 mins
2020-06-17 14:44:42.603643 done after 0.00 mins
2020-06-17 14:44:42.603643 done after 0.00 mins
2020-06-17 14:44:42.611685 done after 0.00 mins
2020-06-17 14:44:42.613719 done after 0.00 mins
2020-06-17 14:44:42.613719 done after 0.00 mins
2020-06-17 14:44:42.623782 done after 0.00 mins
2020-06-17 14:44:42.633883 done after 0.00 mins
2020-06-17 14:44:42.680909 done after 0.00 mins
2020-06-17 14:44:42.693871 done after 0.00 mins
2020-06-17 14:44:42.704840 done after 0.00 mins
2020-06-17 14:44:42.735824 done after 0.

2020-06-17 14:44:47.828544 done after 0.00 mins
2020-06-17 14:44:47.883382 done after 0.00 mins
2020-06-17 14:44:47.940593 done after 0.00 mins
2020-06-17 14:44:48.001042 done after 0.00 mins
2020-06-17 14:44:48.062857 done after 0.00 mins
2020-06-17 14:44:48.090925 done after 0.00 mins
2020-06-17 14:44:48.153368 done after 0.00 mins
2020-06-17 14:44:48.210088 done after 0.00 mins
2020-06-17 14:44:48.272935 done after 0.00 mins
2020-06-17 14:44:48.353691 done after 0.00 mins
2020-06-17 14:44:48.353691 done after 0.00 mins
2020-06-17 14:44:48.363742 done after 0.00 mins
2020-06-17 14:44:48.444506 done after 0.00 mins
2020-06-17 14:44:48.462595 done after 0.00 mins
2020-06-17 14:44:48.504975 done after 0.00 mins
2020-06-17 14:44:48.515083 done after 0.00 mins
2020-06-17 14:44:48.574833 done after 0.00 mins
2020-06-17 14:44:48.585886 done after 0.00 mins
2020-06-17 14:44:48.624768 done after 0.00 mins
2020-06-17 14:44:48.647022 done after 0.00 mins
2020-06-17 14:44:48.708742 done after 0.

2020-06-17 14:44:56.833337 done after 0.00 mins
2020-06-17 14:44:56.843438 done after 0.00 mins
2020-06-17 14:44:56.898647 done after 0.00 mins
2020-06-17 14:44:56.936616 done after 0.00 mins
2020-06-17 14:44:56.953924 done after 0.00 mins
2020-06-17 14:44:57.004763 done after 0.00 mins
2020-06-17 14:44:57.097216 done after 0.00 mins
2020-06-17 14:44:57.175594 done after 0.00 mins
2020-06-17 14:44:57.266032 done after 0.00 mins
2020-06-17 14:44:57.268110 done after 0.00 mins
2020-06-17 14:44:57.268110 done after 0.00 mins
2020-06-17 14:44:57.268110 done after 0.00 mins
2020-06-17 14:44:57.268110 done after 0.00 mins
2020-06-17 14:44:57.278166 done after 0.00 mins
2020-06-17 14:44:57.278166 done after 0.00 mins
2020-06-17 14:44:57.328768 done after 0.00 mins
2020-06-17 14:44:57.336805 done after 0.00 mins
2020-06-17 14:44:57.338840 done after 0.00 mins
2020-06-17 14:44:57.348930 done after 0.00 mins
2020-06-17 14:44:57.358996 done after 0.00 mins
2020-06-17 14:44:57.411724 done after 0.

2020-06-17 14:45:02.326633 done after 0.00 mins
2020-06-17 14:45:02.338729 done after 0.00 mins
2020-06-17 14:45:02.405385 done after 0.00 mins
2020-06-17 14:45:02.478298 done after 0.00 mins
2020-06-17 14:45:02.558889 done after 0.00 mins
2020-06-17 14:45:02.651879 done after 0.00 mins
2020-06-17 14:45:02.733323 done after 0.00 mins
2020-06-17 14:45:02.834518 done after 0.00 mins
2020-06-17 14:45:02.842560 done after 0.00 mins
2020-06-17 14:45:02.844589 done after 0.00 mins
2020-06-17 14:45:02.848165 done after 0.00 mins
2020-06-17 14:45:02.851158 done after 0.00 mins
2020-06-17 14:45:02.854149 done after 0.00 mins
2020-06-17 14:45:02.858138 done after 0.00 mins
2020-06-17 14:45:02.863131 done after 0.00 mins
2020-06-17 14:45:02.887059 done after 0.00 mins
2020-06-17 14:45:02.893044 done after 0.00 mins
2020-06-17 14:45:02.900026 done after 0.00 mins
2020-06-17 14:45:02.907504 done after 0.00 mins
2020-06-17 14:45:02.952272 done after 0.00 mins
2020-06-17 14:45:02.969227 done after 0.

2020-06-17 14:45:08.326152 done after 0.00 mins
2020-06-17 14:45:08.336218 done after 0.00 mins
2020-06-17 14:45:08.402725 done after 0.00 mins
2020-06-17 14:45:08.419681 done after 0.00 mins
2020-06-17 14:45:08.478004 done after 0.00 mins
2020-06-17 14:45:08.528442 done after 0.00 mins
2020-06-17 14:45:08.588910 done after 0.00 mins
2020-06-17 14:45:08.649402 done after 0.00 mins
2020-06-17 14:45:08.733340 done after 0.00 mins
2020-06-17 14:45:08.835496 done after 0.00 mins
2020-06-17 14:45:08.835496 done after 0.00 mins
2020-06-17 14:45:08.835496 done after 0.00 mins
2020-06-17 14:45:08.840030 done after 0.00 mins
2020-06-17 14:45:08.842064 done after 0.00 mins
2020-06-17 14:45:08.842064 done after 0.00 mins
2020-06-17 14:45:08.850299 done after 0.00 mins
2020-06-17 14:45:08.852364 done after 0.00 mins
2020-06-17 14:45:08.852364 done after 0.00 mins
2020-06-17 14:45:08.872689 done after 0.00 mins
2020-06-17 14:45:08.882754 done after 0.00 mins
2020-06-17 14:45:08.890825 done after 0.

2020-06-17 14:45:13.727530 done after 0.00 mins
2020-06-17 14:45:13.812014 done after 0.00 mins
2020-06-17 14:45:13.842932 done after 0.00 mins
2020-06-17 14:45:13.910779 done after 0.00 mins
2020-06-17 14:45:13.969727 done after 0.00 mins
2020-06-17 14:45:14.022278 done after 0.00 mins
2020-06-17 14:45:14.095506 done after 0.00 mins
2020-06-17 14:45:14.096504 done after 0.00 mins
2020-06-17 14:45:14.099494 done after 0.00 mins
2020-06-17 14:45:14.141383 done after 0.00 mins
2020-06-17 14:45:14.145381 done after 0.00 mins
2020-06-17 14:45:14.151355 done after 0.00 mins
2020-06-17 14:45:14.157339 done after 0.00 mins
2020-06-17 14:45:14.165320 done after 0.00 mins
2020-06-17 14:45:14.173460 done after 0.00 mins
2020-06-17 14:45:14.222537 done after 0.00 mins
2020-06-17 14:45:14.234629 done after 0.00 mins
2020-06-17 14:45:14.285017 done after 0.00 mins
2020-06-17 14:45:14.303217 done after 0.00 mins
2020-06-17 14:45:14.354192 done after 0.00 mins
2020-06-17 14:45:14.375876 done after 0.

2020-06-17 14:45:20.005385 done after 0.00 mins
2020-06-17 14:45:20.023517 done after 0.00 mins
2020-06-17 14:45:20.098554 done after 0.00 mins
2020-06-17 14:45:20.137548 done after 0.00 mins
2020-06-17 14:45:20.168000 done after 0.00 mins
2020-06-17 14:45:20.309555 done after 0.00 mins
2020-06-17 14:45:20.388826 done after 0.00 mins
2020-06-17 14:45:20.453268 done after 0.00 mins
2020-06-17 14:45:20.541981 done after 0.00 mins
2020-06-17 14:45:20.544012 done after 0.00 mins
2020-06-17 14:45:20.544012 done after 0.00 mins
2020-06-17 14:45:20.544012 done after 0.00 mins
2020-06-17 14:45:20.555870 done after 0.00 mins
2020-06-17 14:45:20.585881 done after 0.00 mins
2020-06-17 14:45:20.593956 done after 0.00 mins
2020-06-17 14:45:20.604049 done after 0.00 mins
2020-06-17 14:45:20.614111 done after 0.00 mins
2020-06-17 14:45:20.644525 done after 0.00 mins
2020-06-17 14:45:20.654619 done after 0.00 mins
2020-06-17 14:45:20.674767 done after 0.00 mins
2020-06-17 14:45:20.725212 done after 0.

2020-06-17 14:45:24.567088 done after 0.00 mins
2020-06-17 14:45:24.588276 done after 0.00 mins
2020-06-17 14:45:24.637146 done after 0.00 mins
2020-06-17 14:45:24.657368 done after 0.00 mins
2020-06-17 14:45:24.709592 done after 0.00 mins
2020-06-17 14:45:24.758676 done after 0.00 mins
2020-06-17 14:45:24.822864 done after 0.00 mins
2020-06-17 14:45:24.880414 done after 0.00 mins
2020-06-17 14:45:24.909663 done after 0.00 mins
2020-06-17 14:45:24.962071 done after 0.00 mins
2020-06-17 14:45:24.972135 done after 0.00 mins
2020-06-17 14:45:25.009923 done after 0.00 mins
2020-06-17 14:45:25.015906 done after 0.00 mins
2020-06-17 14:45:25.022887 done after 0.00 mins
2020-06-17 14:45:25.030869 done after 0.00 mins
2020-06-17 14:45:25.063178 done after 0.00 mins
2020-06-17 14:45:25.073273 done after 0.00 mins
2020-06-17 14:45:25.083337 done after 0.00 mins
2020-06-17 14:45:25.093844 done after 0.00 mins
2020-06-17 14:45:25.144433 done after 0.00 mins
2020-06-17 14:45:25.162632 done after 0.

2020-06-17 14:45:31.455409 done after 0.00 mins
2020-06-17 14:45:31.506136 done after 0.00 mins
2020-06-17 14:45:31.526459 done after 0.00 mins
2020-06-17 14:45:31.587679 done after 0.00 mins
2020-06-17 14:45:31.658696 done after 0.00 mins
2020-06-17 14:45:31.689433 done after 0.00 mins
2020-06-17 14:45:31.768878 done after 0.00 mins
2020-06-17 14:45:31.848373 done after 0.00 mins
2020-06-17 14:45:31.951618 done after 0.00 mins
2020-06-17 14:45:32.022878 done after 0.00 mins
2020-06-17 14:45:32.128696 done after 0.00 mins
2020-06-17 14:45:32.215552 done after 0.00 mins
2020-06-17 14:45:32.296247 done after 0.00 mins
2020-06-17 14:45:32.364981 done after 0.00 mins
2020-06-17 14:45:32.367007 done after 0.00 mins
2020-06-17 14:45:32.375040 done after 0.00 mins
2020-06-17 14:45:32.377067 done after 0.00 mins
2020-06-17 14:45:32.387177 done after 0.00 mins
2020-06-17 14:45:32.397239 done after 0.00 mins
2020-06-17 14:45:32.449658 done after 0.00 mins
2020-06-17 14:45:32.466614 done after 0.

2020-06-17 14:45:37.906337 done after 0.00 mins
2020-06-17 14:45:37.916415 done after 0.00 mins
2020-06-17 14:45:37.926480 done after 0.00 mins
2020-06-17 14:45:37.989520 done after 0.00 mins
2020-06-17 14:45:38.004480 done after 0.00 mins
2020-06-17 14:45:38.034801 done after 0.00 mins
2020-06-17 14:45:38.052270 done after 0.00 mins
2020-06-17 14:45:38.105232 done after 0.00 mins
2020-06-17 14:45:38.165744 done after 0.00 mins
2020-06-17 14:45:38.188889 done after 0.00 mins
2020-06-17 14:45:38.188889 done after 0.00 mins
2020-06-17 14:45:38.188889 done after 0.00 mins
2020-06-17 14:45:38.229292 done after 0.00 mins
2020-06-17 14:45:38.239358 done after 0.00 mins
2020-06-17 14:45:38.249414 done after 0.00 mins
2020-06-17 14:45:38.257455 done after 0.00 mins
2020-06-17 14:45:38.259494 done after 0.00 mins
2020-06-17 14:45:38.335128 done after 0.00 mins
2020-06-17 14:45:38.345099 done after 0.00 mins
2020-06-17 14:45:38.356072 done after 0.00 mins
2020-06-17 14:45:38.410550 done after 0.

2020-06-17 14:45:44.027973 done after 0.00 mins
2020-06-17 14:45:44.080513 done after 0.00 mins
2020-06-17 14:45:44.121583 done after 0.00 mins
2020-06-17 14:45:44.149815 done after 0.00 mins
2020-06-17 14:45:44.212218 done after 0.00 mins
2020-06-17 14:45:44.281101 done after 0.00 mins
2020-06-17 14:45:44.362414 done after 0.00 mins
2020-06-17 14:45:44.443163 done after 0.00 mins
2020-06-17 14:45:44.503622 done after 0.00 mins
2020-06-17 14:45:44.571215 done after 0.00 mins
2020-06-17 14:45:44.627066 done after 0.00 mins
2020-06-17 14:45:44.679925 done after 0.00 mins
2020-06-17 14:45:44.684914 done after 0.00 mins
2020-06-17 14:45:44.685982 done after 0.00 mins
2020-06-17 14:45:44.696049 done after 0.00 mins
2020-06-17 14:45:44.706125 done after 0.00 mins
2020-06-17 14:45:44.736382 done after 0.00 mins
2020-06-17 14:45:44.756595 done after 0.00 mins
2020-06-17 14:45:44.796767 done after 0.00 mins
2020-06-17 14:45:44.815126 done after 0.00 mins
2020-06-17 14:45:44.861178 done after 0.

2020-06-17 14:45:50.963653 done after 0.00 mins
2020-06-17 14:45:50.963653 done after 0.00 mins
2020-06-17 14:45:50.973728 done after 0.00 mins
2020-06-17 14:45:50.976553 done after 0.00 mins
2020-06-17 14:45:50.980542 done after 0.00 mins
2020-06-17 14:45:50.983534 done after 0.00 mins
2020-06-17 14:45:50.988520 done after 0.00 mins
2020-06-17 14:45:50.994505 done after 0.00 mins
2020-06-17 14:45:51.028414 done after 0.00 mins
2020-06-17 14:45:51.035305 done after 0.00 mins
2020-06-17 14:45:51.035305 done after 0.00 mins
2020-06-17 14:45:51.045378 done after 0.00 mins
2020-06-17 14:45:51.055740 done after 0.00 mins
2020-06-17 14:45:51.096054 done after 0.00 mins
2020-06-17 14:45:51.106129 done after 0.00 mins
2020-06-17 14:45:51.116301 done after 0.00 mins
2020-06-17 14:45:51.126362 done after 0.00 mins
2020-06-17 14:45:51.156794 done after 0.00 mins
2020-06-17 14:45:51.174628 done after 0.00 mins
2020-06-17 14:45:51.237461 done after 0.00 mins
2020-06-17 14:45:51.237461 done after 0.

2020-06-17 14:45:55.432757 done after 0.00 mins
2020-06-17 14:45:55.499173 done after 0.00 mins
2020-06-17 14:45:55.550954 done after 0.00 mins
2020-06-17 14:45:55.601582 done after 0.00 mins
2020-06-17 14:45:55.680074 done after 0.00 mins
2020-06-17 14:45:55.682111 done after 0.00 mins
2020-06-17 14:45:55.690155 done after 0.00 mins
2020-06-17 14:45:55.692183 done after 0.00 mins
2020-06-17 14:45:55.702242 done after 0.00 mins
2020-06-17 14:45:55.742748 done after 0.00 mins
2020-06-17 14:45:55.752878 done after 0.00 mins
2020-06-17 14:45:55.805772 done after 0.00 mins
2020-06-17 14:45:55.818803 done after 0.00 mins
2020-06-17 14:45:55.874247 done after 0.00 mins
2020-06-17 14:45:55.894617 done after 0.00 mins
2020-06-17 14:45:55.944098 done after 0.00 mins
2020-06-17 14:45:56.029048 done after 0.00 mins
2020-06-17 14:45:56.054979 done after 0.00 mins
2020-06-17 14:45:56.108034 done after 0.00 mins
2020-06-17 14:45:56.167004 done after 0.00 mins
2020-06-17 14:45:56.219680 done after 0.

2020-06-17 14:46:04.091462 done after 0.00 mins
2020-06-17 14:46:04.159488 done after 0.00 mins
2020-06-17 14:46:04.232832 done after 0.00 mins
2020-06-17 14:46:04.313815 done after 0.00 mins
2020-06-17 14:46:04.453351 done after 0.00 mins
2020-06-17 14:46:04.504738 done after 0.00 mins
2020-06-17 14:46:04.585632 done after 0.00 mins
2020-06-17 14:46:04.585632 done after 0.00 mins
2020-06-17 14:46:04.585632 done after 0.00 mins
2020-06-17 14:46:04.585632 done after 0.00 mins
2020-06-17 14:46:04.593664 done after 0.00 mins
2020-06-17 14:46:04.593664 done after 0.00 mins
2020-06-17 14:46:04.620993 done after 0.00 mins
2020-06-17 14:46:04.623986 done after 0.00 mins
2020-06-17 14:46:04.626976 done after 0.00 mins
2020-06-17 14:46:04.628972 done after 0.00 mins
2020-06-17 14:46:04.632962 done after 0.00 mins
2020-06-17 14:46:04.636952 done after 0.00 mins
2020-06-17 14:46:04.640940 done after 0.00 mins
2020-06-17 14:46:04.645932 done after 0.00 mins
2020-06-17 14:46:04.650913 done after 0.

2020-06-17 14:46:05.910246 done after 0.00 mins
2020-06-17 14:46:05.920309 done after 0.00 mins
2020-06-17 14:46:05.930384 done after 0.00 mins
2020-06-17 14:46:05.930384 done after 0.00 mins
2020-06-17 14:46:05.930384 done after 0.00 mins
2020-06-17 14:46:05.930384 done after 0.00 mins
2020-06-17 14:46:05.930384 done after 0.00 mins
2020-06-17 14:46:05.930384 done after 0.00 mins
2020-06-17 14:46:05.971187 done after 0.00 mins
2020-06-17 14:46:05.971187 done after 0.00 mins
2020-06-17 14:46:05.977985 done after 0.00 mins
2020-06-17 14:46:05.980976 done after 0.00 mins
2020-06-17 14:46:05.984968 done after 0.00 mins
2020-06-17 14:46:05.989467 done after 0.00 mins
2020-06-17 14:46:05.993457 done after 0.00 mins
2020-06-17 14:46:05.997446 done after 0.00 mins
2020-06-17 14:46:06.001437 done after 0.00 mins
2020-06-17 14:46:06.044321 done after 0.00 mins
2020-06-17 14:46:06.050304 done after 0.00 mins
2020-06-17 14:46:06.058284 done after 0.00 mins
2020-06-17 14:46:06.066262 done after 0.

2020-06-17 14:46:07.513440 done after 0.00 mins
2020-06-17 14:46:07.520427 done after 0.00 mins
2020-06-17 14:46:07.527401 done after 0.00 mins
2020-06-17 14:46:07.533385 done after 0.00 mins
2020-06-17 14:46:07.552411 done after 0.00 mins
2020-06-17 14:46:07.561468 done after 0.00 mins
2020-06-17 14:46:07.579571 done after 0.00 mins
2020-06-17 14:46:07.579571 done after 0.00 mins
2020-06-17 14:46:07.581597 done after 0.00 mins
2020-06-17 14:46:07.581597 done after 0.00 mins
2020-06-17 14:46:07.581597 done after 0.00 mins
2020-06-17 14:46:07.581597 done after 0.00 mins
2020-06-17 14:46:07.589675 done after 0.00 mins
2020-06-17 14:46:07.626369 done after 0.00 mins
2020-06-17 14:46:07.629361 done after 0.00 mins
2020-06-17 14:46:07.634349 done after 0.00 mins
2020-06-17 14:46:07.638336 done after 0.00 mins
2020-06-17 14:46:07.642326 done after 0.00 mins
2020-06-17 14:46:07.646315 done after 0.00 mins
2020-06-17 14:46:07.651301 done after 0.00 mins
2020-06-17 14:46:07.657286 done after 0.

2020-06-17 14:46:09.625641 done after 0.00 mins
2020-06-17 14:46:09.671605 done after 0.00 mins
2020-06-17 14:46:09.679581 done after 0.00 mins
2020-06-17 14:46:09.687560 done after 0.00 mins
2020-06-17 14:46:09.699529 done after 0.00 mins
2020-06-17 14:46:09.744409 done after 0.00 mins
2020-06-17 14:46:09.755377 done after 0.00 mins
2020-06-17 14:46:09.768444 done after 0.00 mins
2020-06-17 14:46:09.816883 done after 0.00 mins
2020-06-17 14:46:09.826960 done after 0.00 mins
2020-06-17 14:46:09.867687 done after 0.00 mins
2020-06-17 14:46:09.875719 done after 0.00 mins
2020-06-17 14:46:09.877746 done after 0.00 mins
2020-06-17 14:46:09.877746 done after 0.00 mins
2020-06-17 14:46:09.877746 done after 0.00 mins
2020-06-17 14:46:09.885812 done after 0.00 mins
2020-06-17 14:46:09.890826 done after 0.00 mins
2020-06-17 14:46:09.894816 done after 0.00 mins
2020-06-17 14:46:09.912767 done after 0.00 mins
2020-06-17 14:46:09.919750 done after 0.00 mins
2020-06-17 14:46:09.926729 done after 0.

2020-06-17 14:46:12.434330 done after 0.00 mins
2020-06-17 14:46:12.443461 done after 0.00 mins
2020-06-17 14:46:12.501962 done after 0.00 mins
2020-06-17 14:46:12.518985 done after 0.00 mins
2020-06-17 14:46:12.522019 done after 0.00 mins
2020-06-17 14:46:12.524048 done after 0.00 mins
2020-06-17 14:46:12.524048 done after 0.00 mins
2020-06-17 14:46:12.524048 done after 0.00 mins
2020-06-17 14:46:12.552267 done after 0.00 mins
2020-06-17 14:46:12.554305 done after 0.00 mins
2020-06-17 14:46:12.554305 done after 0.00 mins
2020-06-17 14:46:12.564371 done after 0.00 mins
2020-06-17 14:46:12.564371 done after 0.00 mins
2020-06-17 14:46:12.574452 done after 0.00 mins
2020-06-17 14:46:12.584535 done after 0.00 mins
2020-06-17 14:46:12.622844 done after 0.00 mins
2020-06-17 14:46:12.624881 done after 0.00 mins
2020-06-17 14:46:12.634989 done after 0.00 mins
2020-06-17 14:46:12.676066 done after 0.00 mins
2020-06-17 14:46:12.686386 done after 0.00 mins
2020-06-17 14:46:12.696560 done after 0.

2020-06-17 14:46:15.172972 done after 0.00 mins
2020-06-17 14:46:15.223330 done after 0.00 mins
2020-06-17 14:46:15.243483 done after 0.00 mins
2020-06-17 14:46:15.324980 done after 0.00 mins
2020-06-17 14:46:15.343411 done after 0.00 mins
2020-06-17 14:46:15.406243 done after 0.00 mins
2020-06-17 14:46:15.424577 done after 0.00 mins
2020-06-17 14:46:15.426614 done after 0.00 mins
2020-06-17 14:46:15.426614 done after 0.00 mins
2020-06-17 14:46:15.426614 done after 0.00 mins
2020-06-17 14:46:15.498319 done after 0.00 mins
2020-06-17 14:46:15.506386 done after 0.00 mins
2020-06-17 14:46:15.508417 done after 0.00 mins
2020-06-17 14:46:15.516453 done after 0.00 mins
2020-06-17 14:46:15.518480 done after 0.00 mins
2020-06-17 14:46:15.528553 done after 0.00 mins
2020-06-17 14:46:15.607860 done after 0.00 mins
2020-06-17 14:46:15.617833 done after 0.00 mins
2020-06-17 14:46:15.626808 done after 0.00 mins
2020-06-17 14:46:15.676675 done after 0.00 mins
2020-06-17 14:46:15.693630 done after 0.

2020-06-17 14:46:19.954513 done after 0.00 mins
2020-06-17 14:46:19.974984 done after 0.00 mins
2020-06-17 14:46:20.009564 done after 0.00 mins
2020-06-17 14:46:20.065974 done after 0.00 mins
2020-06-17 14:46:20.086271 done after 0.00 mins
2020-06-17 14:46:20.148229 done after 0.00 mins
2020-06-17 14:46:20.198640 done after 0.00 mins
2020-06-17 14:46:20.269422 done after 0.00 mins
2020-06-17 14:46:20.269422 done after 0.00 mins
2020-06-17 14:46:20.269422 done after 0.00 mins
2020-06-17 14:46:20.279495 done after 0.00 mins
2020-06-17 14:46:20.279495 done after 0.00 mins
2020-06-17 14:46:20.289556 done after 0.00 mins
2020-06-17 14:46:20.299745 done after 0.00 mins
2020-06-17 14:46:20.351547 done after 0.00 mins
2020-06-17 14:46:20.361520 done after 0.00 mins
2020-06-17 14:46:20.372492 done after 0.00 mins
2020-06-17 14:46:20.409769 done after 0.00 mins
2020-06-17 14:46:20.419820 done after 0.00 mins
2020-06-17 14:46:20.437980 done after 0.00 mins
2020-06-17 14:46:20.470268 done after 0.

2020-06-17 14:46:24.306418 done after 0.00 mins
2020-06-17 14:46:24.316489 done after 0.00 mins
2020-06-17 14:46:24.324578 done after 0.00 mins
2020-06-17 14:46:24.373306 done after 0.00 mins
2020-06-17 14:46:24.384274 done after 0.00 mins
2020-06-17 14:46:24.395253 done after 0.00 mins
2020-06-17 14:46:24.438642 done after 0.00 mins
2020-06-17 14:46:24.457886 done after 0.00 mins
2020-06-17 14:46:24.468037 done after 0.00 mins
2020-06-17 14:46:24.516483 done after 0.00 mins
2020-06-17 14:46:24.518640 done after 0.00 mins
2020-06-17 14:46:24.518640 done after 0.00 mins
2020-06-17 14:46:24.518640 done after 0.00 mins
2020-06-17 14:46:24.528806 done after 0.00 mins
2020-06-17 14:46:24.528806 done after 0.00 mins
2020-06-17 14:46:24.567094 done after 0.00 mins
2020-06-17 14:46:24.569245 done after 0.00 mins
2020-06-17 14:46:24.579352 done after 0.00 mins
2020-06-17 14:46:24.593381 done after 0.00 mins
2020-06-17 14:46:24.637625 done after 0.00 mins
2020-06-17 14:46:24.647990 done after 0.

2020-06-17 14:46:29.133603 done after 0.00 mins
2020-06-17 14:46:29.153764 done after 0.00 mins
2020-06-17 14:46:29.214344 done after 0.00 mins
2020-06-17 14:46:29.305100 done after 0.00 mins
2020-06-17 14:46:29.325275 done after 0.00 mins
2020-06-17 14:46:29.441466 done after 0.00 mins
2020-06-17 14:46:29.496318 done after 0.00 mins
2020-06-17 14:46:29.559542 done after 0.00 mins
2020-06-17 14:46:29.628006 done after 0.00 mins
2020-06-17 14:46:29.698924 done after 0.00 mins
2020-06-17 14:46:29.767523 done after 0.00 mins
2020-06-17 14:46:29.772547 done after 0.00 mins
2020-06-17 14:46:29.776537 done after 0.00 mins
2020-06-17 14:46:29.782521 done after 0.00 mins
2020-06-17 14:46:29.791496 done after 0.00 mins
2020-06-17 14:46:29.800471 done after 0.00 mins
2020-06-17 14:46:29.840006 done after 0.00 mins
2020-06-17 14:46:29.850452 done after 0.00 mins
2020-06-17 14:46:29.910154 done after 0.00 mins
2020-06-17 14:46:29.927110 done after 0.00 mins
2020-06-17 14:46:29.972794 done after 0.

2020-06-17 14:46:37.517759 done after 0.00 mins
2020-06-17 14:46:37.527863 done after 0.00 mins
2020-06-17 14:46:37.527863 done after 0.00 mins
2020-06-17 14:46:37.527863 done after 0.00 mins
2020-06-17 14:46:37.565170 done after 0.00 mins
2020-06-17 14:46:37.569158 done after 0.00 mins
2020-06-17 14:46:37.572150 done after 0.00 mins
2020-06-17 14:46:37.576140 done after 0.00 mins
2020-06-17 14:46:37.580129 done after 0.00 mins
2020-06-17 14:46:37.584118 done after 0.00 mins
2020-06-17 14:46:37.588095 done after 0.00 mins
2020-06-17 14:46:37.596128 done after 0.00 mins
2020-06-17 14:46:37.633520 done after 0.00 mins
2020-06-17 14:46:37.639044 done after 0.00 mins
2020-06-17 14:46:37.649110 done after 0.00 mins
2020-06-17 14:46:37.659182 done after 0.00 mins
2020-06-17 14:46:37.659182 done after 0.00 mins
2020-06-17 14:46:37.679325 done after 0.00 mins
2020-06-17 14:46:37.689398 done after 0.00 mins
2020-06-17 14:46:37.699497 done after 0.00 mins
2020-06-17 14:46:37.709564 done after 0.

2020-06-17 14:46:41.712874 done after 0.00 mins
2020-06-17 14:46:41.769662 done after 0.00 mins
2020-06-17 14:46:41.797886 done after 0.00 mins
2020-06-17 14:46:41.854663 done after 0.00 mins
2020-06-17 14:46:41.910664 done after 0.00 mins
2020-06-17 14:46:41.961403 done after 0.00 mins
2020-06-17 14:46:42.022263 done after 0.00 mins
2020-06-17 14:46:42.097164 done after 0.00 mins
2020-06-17 14:46:42.101154 done after 0.00 mins
2020-06-17 14:46:42.107137 done after 0.00 mins
2020-06-17 14:46:42.113120 done after 0.00 mins
2020-06-17 14:46:42.122098 done after 0.00 mins
2020-06-17 14:46:42.130078 done after 0.00 mins
2020-06-17 14:46:42.183932 done after 0.00 mins
2020-06-17 14:46:42.197894 done after 0.00 mins
2020-06-17 14:46:42.253260 done after 0.00 mins
2020-06-17 14:46:42.272210 done after 0.00 mins
2020-06-17 14:46:42.354195 done after 0.00 mins
2020-06-17 14:46:42.374442 done after 0.00 mins
2020-06-17 14:46:42.473759 done after 0.00 mins
2020-06-17 14:46:42.563059 done after 0.

2020-06-17 14:46:49.701316 done after 0.00 mins
2020-06-17 14:46:49.766142 done after 0.00 mins
2020-06-17 14:46:49.791075 done after 0.00 mins
2020-06-17 14:46:49.896227 done after 0.00 mins
2020-06-17 14:46:49.960845 done after 0.00 mins
2020-06-17 14:46:50.035648 done after 0.00 mins
2020-06-17 14:46:50.114856 done after 0.00 mins
2020-06-17 14:46:50.199632 done after 0.00 mins
2020-06-17 14:46:50.289390 done after 0.00 mins
2020-06-17 14:46:50.371487 done after 0.00 mins
2020-06-17 14:46:50.468944 done after 0.00 mins
2020-06-17 14:46:50.554715 done after 0.00 mins
2020-06-17 14:46:50.562692 done after 0.00 mins
2020-06-17 14:46:50.631673 done after 0.00 mins
2020-06-17 14:46:50.640728 done after 0.00 mins
2020-06-17 14:46:50.652994 done after 0.00 mins
2020-06-17 14:46:50.720815 done after 0.00 mins
2020-06-17 14:46:50.761872 done after 0.00 mins
2020-06-17 14:46:50.825182 done after 0.00 mins
2020-06-17 14:46:50.845936 done after 0.00 mins
2020-06-17 14:46:50.914800 done after 0.

**Note**

Before parallization. Total runs for estimated 1,000,000 loops took about 3 min

* 2020-06-17 14:41:28.739905 (Beginning!)

* 2020-06-17 14:46:51.676526 (End!)

**Note**

Try to change the above code snippet to run parallelization.

Please do not use jupyter notbook, it won't reflect anything.

If you are like me, curious about strategy's mean-reversion velocity, you might want to read this research report.

[Critical-Line Algorithm for Portfolio Optimization by Bailey and Lopez [2013]](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2197616)

**Hint**

Vectorize iterable then try using "iterrows()" or "itertuples()" if you are using pandas.

In [6]:
print("Total count provided: {0}, total count retrieved: {1}".format(len(jobs), len(df)))
df

Total count provided: 10000, total count retrieved: 10000


[[{'profit': 0.5,
   'loss': 0.5,
   'mean': 1.6614332029031802,
   'std': 0.971519953603021,
   'sharpe': 1.7101380128545143}],
 [{'profit': 0.5,
   'loss': 1.0,
   'mean': 1.690836960196075,
   'std': 0.8553741629691958,
   'sharpe': 1.976722039775904}],
 [{'profit': 0.5,
   'loss': 1.5,
   'mean': 1.6019704310539382,
   'std': 0.665186548309494,
   'sharpe': 2.4083025057033822}],
 [{'profit': 0.5,
   'loss': 2.0,
   'mean': 1.6800109553709877,
   'std': 0.7956590529266577,
   'sharpe': 2.111470923621663}],
 [{'profit': 0.5,
   'loss': 2.5,
   'mean': 1.6284748547761916,
   'std': 0.7573222142544277,
   'sharpe': 2.1503064668179586}],
 [{'profit': 0.5,
   'loss': 3.0,
   'mean': 1.6916628779460277,
   'std': 0.7682696911585473,
   'sharpe': 2.201912814489672}],
 [{'profit': 0.5,
   'loss': 3.5,
   'mean': 1.6814481655854396,
   'std': 0.7648066514547471,
   'sharpe': 2.1985271210535875}],
 [{'profit': 0.5,
   'loss': 4.0,
   'mean': 1.6920193715068808,
   'std': 0.7769855350757856,
 

In [7]:
df1 = rs.opt_tr(profit_range = profit_range,
                  loss_range = loss_range,
                  sigma = sigma,
                  n_run = n_run,
                  max_period = max_period,
                  forecast = forecast, 
                  half_life = half_life,
                  num_threads = 3)

[      profit  loss      mean       std    sharpe
0        0.5   0.5  1.396264  0.984278  1.418566
1        0.5   1.0  1.587675  0.774706  2.049392
2        0.5   1.5  1.712608  0.826244  2.072764
3        0.5   2.0  1.600266  0.662226  2.416496
4        0.5   2.5  1.610815  0.680407  2.367427
...      ...   ...       ...       ...       ...
3329     3.5   5.0  1.380343  4.395407  0.314042
3330     3.5   5.5  2.112414  3.886711  0.543497
3331     3.5   6.0  2.024534  4.154450  0.487317
3332     3.5   6.5  1.968073  4.356944  0.451709
3333     3.5   7.0  3.004285  3.233671  0.929063

[3334 rows x 5 columns]] this out


[      profit  loss      mean       std    sharpe
0        0.5   0.5  1.396264  0.984278  1.418566
1        0.5   1.0  1.587675  0.774706  2.049392
2        0.5   1.5  1.712608  0.826244  2.072764
3        0.5   2.0  1.600266  0.662226  2.416496
4        0.5   2.5  1.610815  0.680407  2.367427
...      ...   ...       ...       ...       ...
3329     3.5   5.0  1.380343  4.395407  0.314042
3330     3.5   5.5  2.112414  3.886711  0.543497
3331     3.5   6.0  2.024534  4.154450  0.487317
3332     3.5   6.5  1.968073  4.356944  0.451709
3333     3.5   7.0  3.004285  3.233671  0.929063

[3334 rows x 5 columns],       profit  loss      mean       std     sharpe
6667     7.0   4.0 -4.659314  0.477871  -9.750155
6668     7.0   4.5 -5.038846  0.450071 -11.195665
6669     7.0   5.0 -5.680721  0.604291  -9.400631
6670     7.0   5.5 -6.030476  0.483595 -12.470101
6671     7.0   6.0 -6.486064  0.876892  -7.396649
...      ...   ...       ...       ...        ...
9995    10.0   8.0 -3.843756  6.173

[      profit  loss      mean       std    sharpe
0        0.5   0.5  1.396264  0.984278  1.418566
1        0.5   1.0  1.587675  0.774706  2.049392
2        0.5   1.5  1.712608  0.826244  2.072764
3        0.5   2.0  1.600266  0.662226  2.416496
4        0.5   2.5  1.610815  0.680407  2.367427
...      ...   ...       ...       ...       ...
3329     3.5   5.0  1.380343  4.395407  0.314042
3330     3.5   5.5  2.112414  3.886711  0.543497
3331     3.5   6.0  2.024534  4.154450  0.487317
3332     3.5   6.5  1.968073  4.356944  0.451709
3333     3.5   7.0  3.004285  3.233671  0.929063

[3334 rows x 5 columns],       profit  loss      mean       std     sharpe
6667     7.0   4.0 -4.659314  0.477871  -9.750155
6668     7.0   4.5 -5.038846  0.450071 -11.195665
6669     7.0   5.0 -5.680721  0.604291  -9.400631
6670     7.0   5.5 -6.030476  0.483595 -12.470101
6671     7.0   6.0 -6.486064  0.876892  -7.396649
...      ...   ...       ...       ...        ...
9995    10.0   8.0 -3.843756  6.173

2020-06-17 14:49:46.331662 100.0% _opt_tr done after 2.88 mins. Remaining 0.0 mins.


In [8]:
df1

profit  loss      mean       std    sharpe
0        0.5   0.5  1.396264  0.984278  1.418566
1        0.5   1.0  1.587675  0.774706  2.049392
2        0.5   1.5  1.712608  0.826244  2.072764
3        0.5   2.0  1.600266  0.662226  2.416496
4        0.5   2.5  1.610815  0.680407  2.367427
...      ...   ...       ...       ...       ...
9995    10.0   8.0 -3.843756  6.173427 -0.622629
9996    10.0   8.5 -4.054103  7.336595 -0.552587
9997    10.0   9.0 -4.632491  6.437791 -0.719578
9998    10.0   9.5 -4.404237  7.476366 -0.589088
9999    10.0  10.0 -4.952120  6.615454 -0.748568

[10000 rows x 5 columns]

**Conclusion**

The above was ran parallel with 3 threads:
    
1. 2020-06-17 14:49:16.186467
2. 2020-06-17 14:49:21.045204
3. 2020-06-17 14:49:46.331662

Based on the time provided by python system, it seems that it took less than 40 seconds.

When running large dataset, especially with Pandas parallelization can save time and always vectorize the datasets.

**Note**

If the answer to the above will not tally with the initial run (Scalar Operations), since we are running a Monte-Carlos.